### ISSUE TITLE
Enable server-side config validation with webhooks

### ISSUE THREAD

**Issue by ayj** [istio/istio#1894]
Pilot and Mixer components can perform server-side validation of configuration through k8s webhooks. This code has already be implemented but has not yet been enabled by default due to cloud provider [bug](https://github.com/kubernetes/kubernetes/issues/49987).

TODO
- [x] Update Istio install script to generate and install certs for Pilot and Mixer webhook validation services. 
- [ ] Update Pilot e2e test to use validation webhook by default on k8s >= 1.9
- [ ] Update Pilot validation webhook to upgrade to latest install recipe
- [ ] Update Mixer e2e test to use validation webhook by default on k8s >= 1.9
- [ ] Update Pilot validation webhook to upgrade to latest install recipe
- [ ] Documentation on istio.io under kubernetes specific setups, and config concepts.  

After this is done, the feature is always on.  We can consider it to be beta.

**Comment by cmluciano** [comment]
I think we have a few of these completed. Can we link and check off the boxes?

**Comment by ayj** [comment]
@jmuk, the pilot specific portions of this issue are completed. I've also updated e2e tests to use GKE >= 1.8.x so that validation can be enabled by default. You should be able to use the same cert generation script for mixer webhook service. 

https://istio.io/docs/setup/kubernetes/quick-start.html still needs to be updated to include steps for cert generation.

Similar updates for webhook cert provisioning are required for any other platform variant, e.g. GKE (@salrashid123, @selmanj), Consul (@GregHanson), and Eureka (@ijsnellf).

**Comment by xiaolanz** [comment]
@ayj Is this work item using https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation

**Comment by ayj** [comment]
No, we're using ValidatingWebhooks for both schema and semantic validation. We could make use of that feature with some combination of (1) using protobuf annotations to describe constraints, (2) generate corresponding openAPIv3Schema, and (3) providing that in CRD definition.

**Comment by ozevren** [comment]
Folks, can I ask for a favor? When doing this work, can you implement it so that it is a stand-alone library that doesn't have direct dependencies to Pilot, Mixer? So that we can, move it easily in the future (i.e. maybe, to Galley?)

Thanks

**Comment by ayj** [comment]
The Pilot config validation webhook is already mostly decoupled from Pilot except for the CRD and configuration packages.

```golang
	"istio.io/istio/pilot/pkg/config/kube/crd"
	"istio.io/istio/pilot/pkg/model"
```

`crd` would be replaced by whatever common config client we end up introducing. `model` is required to convert to the type-specific objects and for validation.go. The latter part could be mostly replaced by CRD openAPIv3Schema validation.

I have no clue how dependent (or not) Mixer validation logic is to Mixer proper. cc @jmuk

**Comment by yusuoh** [comment]
It seems that the pilot validation logic is only enabled in e2e test. Do we want to enable it in the normal installation?

**Comment by ayj** [comment]
We should. It requires the same cert provisioning steps used by the injection webook for installation.

**Comment by xiaolanz** [comment]
One question is that what's the minimal k8s version Istio will support? Do we allow beta server side validation feature that only works for 1.9 to be on with 1.8 k8s?

**Comment by xiaolanz** [comment]
other click stops are not done yet.

**Comment by ozevren** [comment]
Enabled by default with https://github.com/istio/istio/pull/6035
